rs/pocket_ic_server/src/pocket_ic.rs

```diff
@@ -831,6 +843,233 @@ impl Operation for GetTime {
+) -> OpOut {
+    let subnet_id =
+        ic_types::SubnetId::new(ic_types::PrincipalId(mock_canister_http_response.subnet_id));
+    let subnet = pic.get_subnet_with_id(subnet_id);
+    match subnet {
+        Some(subnet) => {
+            let canister_http_request_id =
+                CanisterHttpRequestId::from(mock_canister_http_response.request_id);
+            let contexts = subnet.canister_http_request_contexts();
+            if let Some(context) = contexts.get(&canister_http_request_id) {
+                let timeout = context.time + Duration::from_secs(5 * 60);
+                let canister_id = context.request.sender;
+                let content = match &mock_canister_http_response.response {
+                    CanisterHttpResponse::CanisterHttpReply(reply) => {
+                        let grpc_channel = pic.runtime.block_on(setup_adapter_mock(Ok(
+                            CanisterHttpSendResponse {
+                                status: reply.status.into(),
+                                headers: reply
+                                    .headers
+                                    .iter()
+                                    .map(|h| ic_https_outcalls_service::HttpHeader {
+                                        name: h.name.clone(),
+                                        value: h.value.clone(),
+                                    })
+                                    .collect(),
+                                content: reply.body.clone(),
+                            },
+                        )));
+                        let query_handler = subnet.query_handler.clone();
+                        let query_handler = BoxCloneService::new(service_fn(move |arg| {
+                            let query_handler = query_handler.clone();
+                            async {
+                                let r = query_handler
+                                    .oneshot(arg)
+                                    .await
+                                    .expect("Inner service should be alive. I hope.");
+                                Ok(r)
+                            }
+                        }));
+                        let mut client = CanisterHttpAdapterClientImpl::new(
+                            pic.runtime.handle().clone(),
+                            grpc_channel,
+                            query_handler.clone(),
+                            1,
+                            MetricsRegistry::new(),
+                            subnet.get_subnet_type(),
+                        );
+                        client
+                            .send(ic_types::canister_http::CanisterHttpRequest {
+                                timeout,
+                                id: canister_http_request_id,
+                                context: context.clone(),
+                            })
+                            .unwrap();
+                        let response = loop {
```

Comment:
There is no timeout here. (Why) can we be confident that the client will respond eventually?

Reply:
Good point! It would only not respond if there was a bug in production code. But then the https outcalls would get stuck in production, too, as there's no timeout in production either (see [here](https://github.com/dfinity/ic/blob/3a31b54c34e74e4a1f541627b6d3b0fbda14b6a0/rs/https_outcalls/consensus/src/pool_manager.rs#L235) - although there we break out of the loop `Err(TryReceiveError::Empty) => break`, if we can't receive anything, then we don't make any progress with https outcalls either). @tthebst Could you please confirm that it's safe to not have a timeout here?